### ISSUE TITLE
Rasterize HTML, PDFs to Raw

### ISSUE THREAD

**Issue by tresf** [qzind/tray#19]
After supporting the 2.0 branch for a couple months, we've noticed an increasing amount of people trying to use Pixel printing against their raw devices for some specific cases, such as a packing slip (something that FedEx or UPS may not provide in raw format) or printing continuous feed image data (HTML) to a receipt printer.

Today, they're trying to print using the 2d printing (HTML, etc), but this often requires them to utilize a different device driver and worse, actually yields different results versus just using raw ZPL, EPL, ESCP, etc.

This enhancement is to allow raw printing to use Pixel features such as HTML, PDF, but then wrap them to the raw equivalent.

![image](https://cloud.githubusercontent.com/assets/6345473/14086241/23ed8acc-f4f2-11e5-9d02-25b7bf621c02.png)

For example, today this is perfectly legal:

```javascript
var data = 
['^XA', { 
   type: 'raw', format: 'image', data: 'assets/img/image_sample_bw.png', 
   options: { language: "ZPL" } 
}, '^XZ'];
```

But this is not... 

```javascript
var data = 
['^XA', { 
   type: 'raw', format: 'pdf', data: 'assets/img/pdf_sample.pdf', 
   options: { language: "ZPL" } 
}, '^XZ'];
```

Since we have the ability to rasterize PDF and HTML, there's no reason why we can't support these formats for raw printers too by rendering them, treating them as raster graphics, and then converting them to raw commands.

**Comment by bberenz** [comment]
Do we need to subset these new formats? As pdf, html, and image each have their own formats when used independantly. And this has even led to some confusion already, causing #70.

**Comment by tresf** [comment]
From our discussion, this will be an expansion of the API.

```javascript
var data = 
['^XA', { 
   type: 'raw', format: 'image', flavor: 'file', data: 'assets/img/image_sample_bw.png', 
   options: { language: "ZPL" } 
}, '^XZ'];
```

**Comment by bberenz** [comment]
Feature added in 9c39650b9b126400d03f10a140da8c753cc24b53
Everything looks right when printing to a file, but you'll need to verify for me on a physical device.

**Comment by tresf** [comment]
> Everything looks right when printing to a file, but you'll need to verify for me on a physical device.

Looks pretty good.  In the case of the PDF, it only takes up about 2/3s of the height and width of the paper.  How do you suggest we specify PDF size?

HTML I had a bit more control over due to the `pageWidth` parameter.

**Comment by bberenz** [comment]
> How do you suggest we specify PDF size?

Do you want to extend the `pageWidth` and `pageHeight` options to include pdf as well?

> Also, I think it may be time to add a radio button for the language type 

32ed9f1b0035eb68c229c4b59608d92cf9c29b2f

**Comment by tresf** [comment]
> Do you want to extend the pageWidth and pageHeight options to include pdf as well?

Sounds like a good plan.

**Comment by bberenz** [comment]
Adjusted with 5e7afd02e8ab30cd57877dd0f1984a9121ebcd88

**Comment by tresf** [comment]
For the new sample page, let's specify the following:
```
EPL2    ZPLII
ESCP    EPCL
```

ESCP is the one I'm not too sure about.  I know we use `ESC *` technique which is clearly defined in the ESCP2 manual, but I can't find much reference to ESCP1 anywhere, perhaps @lite1979 can dig up more information.

Also, no need to show `Image X` and `Image Y` for anything except EPL2 (or CPCL, if it gets added eventually).

**Comment by tresf** [comment]
Some testing:
 * For our 8.5 x 11 `pdf_sample.pdf` sent via EPL2 render width of `812` to the Epson printer, this feature works great
 * When testing the `fourbysix.pdf`, render width of `812` it prints very small in upper right corner.

Any idea why the smaller PDF is printed so small?

```
[DEBUG] 2016-08-12 01:42:03,365 @ qz.printer.PrintServiceMatcher:?
	Searching for PrintService matching raw
[DEBUG] 2016-08-12 01:42:03,365 @ qz.printer.PrintServiceMatcher:?
	Found 4 printers
[DEBUG] 2016-08-12 01:42:03,383 @ qz.printer.PrintServiceMatcher:?
	Found match: Zebra LP2844 Raw
[DEBUG] 2016-08-12 01:42:31,599 @ qz.ws.PrintSocketClient:?
	Message: {"call":"print","promise":{},"params":{"printer":{"name":"Zebra LP2844 Raw"},"options":{"colorType":"color","copies":"1","density":"","duplex":false,"fallbackDensity":600,"interpolation":"","jobName":"","margins":"0","orientation":"","paperThickness":"","printerTray":"","rasterize":false,"rotation":"0","scaleContent":true,"size":null,"units":"in","altPrinting":false,"encoding":"","endOfDoc":"","perSpool":"1"},"data":["\nN\n",{"type":"raw","format":"pdf","flavor":"file","data":"file:///Users/tres/tray/assets/pdf_sample.pdf","options":{"language":"EPL","x":"0","y":"0","dotDensity":"single","xmlTag":"v7:Image","pageWidth":"812","pageHeight":""}},"\nP1,1\n"]},"timestamp":1470980551597,"uid":"9uzftj"}
[WARN] 2016-08-12 01:42:31,600 @ qz.ws.PrintSocketClient:?
	Bad signature on request
[INFO] 2016-08-12 01:42:32,375 @ qz.common.TrayManager:?
	Allowed localhost to print to Zebra LP2844 Raw
[TRACE] 2016-08-12 01:42:32,375 @ qz.utils.PrintingUtilities:?
	Waiting for processor, 0/7 already in use
[DEBUG] 2016-08-12 01:42:32,376 @ qz.utils.PrintingUtilities:?
	Using qz.printer.action.PrintRaw to print
[DEBUG] 2016-08-12 01:42:32,376 @ qz.printer.PrintServiceMatcher:?
	Searching for PrintService matching Zebra LP2844 Raw
[DEBUG] 2016-08-12 01:42:32,376 @ qz.printer.PrintServiceMatcher:?
	Found 4 printers
[DEBUG] 2016-08-12 01:42:32,393 @ qz.printer.PrintServiceMatcher:?
	Found match: Zebra LP2844 Raw
[DEBUG] 2016-08-12 01:42:32,395 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, ZDesigner_LP2844, -l]
[DEBUG] 2016-08-12 01:42:32,679 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS ZDesigner_LP2844: 203dpi
[DEBUG] 2016-08-12 01:42:32,679 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, Zebra_LP2844_Raw, -l]
[DEBUG] 2016-08-12 01:42:32,698 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, PDFwriter, -l]
[DEBUG] 2016-08-12 01:42:32,717 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS PDFwriter: 300dpi
[DEBUG] 2016-08-12 01:42:32,717 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, EpsonTM88IV, -l]
[INFO] 2016-08-12 01:42:33,361 @ qz.printer.ImageWrapper:?
	Loading BufferedImage
[INFO] 2016-08-12 01:42:33,362 @ qz.printer.ImageWrapper:?
	Dimensions: 288x432
[INFO] 2016-08-12 01:42:33,362 @ qz.printer.ImageWrapper:?
	Initializing Image Fields
[INFO] 2016-08-12 01:42:33,362 @ qz.printer.ImageWrapper:?
	Converting image to monochrome
[INFO] 2016-08-12 01:42:33,365 @ qz.printer.ImageWrapper:?
	Packing bits
[INFO] 2016-08-12 01:42:33,366 @ qz.printer.ImageWrapper:?
	Generating byte array
[TRACE] 2016-08-12 01:42:33,369 @ qz.printer.action.PrintRaw:?
	Sending print job to printer
[DEBUG] 2016-08-12 01:42:33,371 @ qz.printer.action.PrintRaw:?
	PrintEvent on sun.print.UnixPrintJob@47de58a7
[DEBUG] 2016-08-12 01:42:33,676 @ qz.printer.action.PrintRaw:?
	PrintEvent on sun.print.UnixPrintJob@47de58a7
[DEBUG] 2016-08-12 01:42:33,677 @ qz.printer.action.PrintRaw:?
	PrintEvent on sun.print.UnixPrintJob@47de58a7
[DEBUG] 2016-08-12 01:42:33,677 @ qz.printer.action.PrintRaw:?
	PrintEvent on sun.print.UnixPrintJob@47de58a7
[TRACE] 2016-08-12 01:42:33,677 @ qz.printer.action.PrintRaw:?
	Print job received by printer
[INFO] 2016-08-12 01:42:33,677 @ qz.utils.PrintingUtilities:?
	Printing complete
[TRACE] 2016-08-12 01:42:33,677 @ qz.utils.PrintingUtilities:?
	Returning processor back to pool
```

**Comment by bberenz** [comment]
Looks like I never had the pdf scale when the size was set..
1d4e577444484eb77b09aebffdf94139519ab7e4 should fix your issue.

**Comment by tresf** [comment]
@bberenz is there a good reason why the pageWidth and pageHeight are in pixels rather than units?

**Comment by bberenz** [comment]
Because we don't take in a `units` option for raw printing.

**Comment by tresf** [comment]
So, I've tested this.

 * Against 203 dpi Zebra printer, pageWidth of `812` and results are pretty good.  Here's a side-by-side of desired and actual.
   ![image](https://cloud.githubusercontent.com/assets/6345473/17638790/94bf8e56-60ba-11e6-8512-41a25dffc72a.png)

The printable area seems to be slightly under 812 (203 * 4), so adjusting this to 800 seems to fix this in the right direction.

 * Against 180 dpi Epson receipt printer, pageWidth of `562.5` and results are dead on (or as close as they're going to get).

This looks good.  We can open separate bug reports as needed.  Closed via 9c39650 5e7afd0.  This feature will be available in 2.1.
